src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -168,21 +171,103 @@ trait EntityComponent {
       allEntitiesAction.flatMap(cloneEntities(destWorkspaceContext, _))
     }
 
+    def insertEntitiesSql(destWorkspaceContext: SlickWorkspaceContext, entities: Map[Entity, UUID]) = {
+      val workspaceId = destWorkspaceContext.workspaceId
+
+      val insertBatches = entities.map { case (entity, entityId) =>
+        sql"($entityId, ${entity.name}, ${entity.entityType}, $workspaceId)"
+      }.grouped(1000).toSeq
+
+      DBIO.seq(insertBatches.map { insertBatch =>
+        val prefix = sql"insert into ENTITY (id, name, entity_type, workspace_id) values "
+        val suffix = insertBatch.reduce { (a, b) =>
+          concatSqlActionsWithDelim(a, b, ",")
+        }
+        concatSqlActions(prefix, suffix).as[Int]
+      }:_*)
+    }
+
+    def insertAttributesSql(destWorkspaceContext: SlickWorkspaceContext, attrsWithEntityIdMap: Seq[(UUID, (String, Attribute), UUID, Option[Int])]) = {
+      val insertBatches = attrsWithEntityIdMap.map { case (attrId, attr, entityId, listIndex) =>
+        val name = attr._1
+        attr._2 match {
+          case value: AttributeValue =>
+            value match {
+              case AttributeBoolean(value) =>
+                sql"($attrId, $name, NULL, NULL, $value, NULL, NULL)"
+              case AttributeNumber(b) =>
+                val value = b.toDouble
+                sql"($attrId, $name, NULL, $value, NULL, NULL, NULL)"
+              case AttributeString(value) =>
+                sql"($attrId, $name, $value, NULL, NULL, NULL, NULL)"
+              case AttributeNull =>
+                sql"($attrId, $name, NULL, NULL, NULL, NULL, NULL)"
+              case _ => throw new RawlsException(s"Unable to save attribute [${value}]")
+            }
+          case ref: AttributeEntityReference =>
+            listIndex match {
+              case Some(index) =>
+                sql"($attrId, $name, NULL, NULL, NULL, (select id from ENTITY where name=${ref.entityName} and entity_type=${ref.entityType} and workspace_id=${destWorkspaceContext.workspaceId}), $index)"
```

Comment:
I think it might be faster to pre-fetch all the referenced entities and populate the ids directly instead of these sub selects